File: 371_Sum_of_Two_Integers.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/sum-of-two-integers/
# Calculate the sum of two integers a and b, but you are not allowed to use the operator + and -.

# Use mask to isolate 32 bits. Sum of 2 bits is the XOR. If both bits are set, carry to next bit. Repeat until no
# more carry. Ir result is more than MAX_INT, subtract 2**32.
# Time - O(log n)
# Space - O(1)

class Solution(object):
    def getSum(self, a, b):
        """
        :type a: int
        :type b: int
        :rtype: int
        """
        MASK = 0xFFFFFFFF       # 2**32 - 1
        MAX_INT = 0x7FFFFFFF    # 2**31 - 1

        while b != 0:

            total = (a ^ b) & MASK
            carry = ((a & b) << 1) & MASK   # shift pattern when both bits are set
            print(a, b, total, carry)
            a, b = total, carry

        return a if a < MAX_INT else ~(a ^ MASK)    # a - 2**321

File: java/371_Sum_of_Two_Integers.java
/*
https://leetcode.com/problems/sum-of-two-integers/
Calculate the sum of two integers a and b, but you are not allowed to use the operator + and -.

Sum without carry is a XOR b.  Carry is when same bit is set in a AND b, shifted left.  Repeat until no carry.
Time - O(1), since 32 bits
Space - O(1)
*/

public class Solution {
    public int getSum(int a, int b) {
        while (b != 0) {
            int sum = a ^ b;
            int carry = (a & b) << 1;
            a = sum;
            b = carry;
        }
        return a;
    }
}
